packages/mds-geography-author/api.ts

```diff
@@ -0,0 +1,292 @@
+    }),
+    async (req, res) => {
+      const { geography_id } = req.params
+      try {
+        const geography = await db.readSingleGeography(geography_id)
+        if (!geography.publish_date && !res.locals.scopes.includes('geographies:read:unpublished')) {
+          throw new InsufficientPermissionsError('permission to read unpublished geographies missing')
+        } else {
+          return res.status(200).send(geography)
+        }
+      } catch (err) {
+        await log.error('failed to read geography', err.stack)
+        if (err instanceof NotFoundError) {
+          return res.status(404).send({ error: err })
+        }
+
+        if (err instanceof InsufficientPermissionsError) {
+          return res.status(403).send({ error: err })
+        }
+
+        return res.status(500).send({ error: new ServerError() })
+      }
+    }
+  )
+
+  app.post(
+    pathsFor('/geographies/'),
+    checkAccess(scopes => scopes.includes('geographies:write')),
+    async (req, res) => {
+      const geography = req.body
+
+      try {
+        const details = geographyValidationDetails(geography)
+        if (details) {
+          throw new ValidationError(JSON.stringify(details))
+        }
+
+        const recorded_geography = await db.writeGeography(geography)
+        return res.status(201).send(recorded_geography)
+      } catch (err) {
+        await log.error('POST /geographies failed', err.stack)
+        if (err.code === '23505') {
+          return res
+            .status(409)
+            .send({ result: `geography ${geography.geography_id} already exists! Did you mean to PUT?` })
+        }
+        if (err instanceof ValidationError) {
+          return res.status(400).send({ error: err })
+        }
+        /* istanbul ignore next */
+        return res.status(500).send({ error: err })
+      }
+    }
+  )
+
+  app.put(
+    pathsFor('/geographies/:geography_id'),
+    checkAccess(scopes => scopes.includes('geographies:write')),
+    async (req, res) => {
+      const geography = req.body
+      try {
+        const details = geographyValidationDetails(geography)
+        if (details) {
+          throw new ValidationError(JSON.stringify(details))
+        }
+        await db.editGeography(geography)
+        return res.status(201).send(geography)
+      } catch (err) {
+        await log.error('failed to edit geography', err.stack)
+        if (err instanceof NotFoundError) {
+          return res.status(404).send({ result: 'not found' })
+        }
+        if (err instanceof ValidationError) {
+          return res.status(400).send({ error: err })
+        }
+        return res.status(500).send(new ServerError(err))
+      }
+    }
+  )
+
+  app.delete(
+    pathsFor('/geographies/:geography_id'),
+    checkAccess(scopes => scopes.includes('geographies:write')),
+    async (req, res) => {
+      const { geography_id } = req.params
+      try {
+        await db.deleteGeographyMetadata(geography_id)
```

Comment:
If we change `deleteGeographyMetadata()` to `throw` if there's no associated `geography`, we could parallelize
```
await Promise.all([db.deleteGeographyMetadata(geography_id), db.deleteGeography(geography_id)])
```

Reply:
Isn't there the risk of a race where the geography and metadata are both present, but the geography gets deleted first, and then the metadata doesn't get deleted?

Reply:
I didn't consider that case, that sounds like it would cause some issues.